crux/src/Crux/Report.hs

```diff
@@ -104,83 +107,119 @@ renderSideConds opts = concatMapM (go [])
   flatBranch []                  = []
 
   isGoal x = case x of
-               Goal {} -> True
+               ProvedGoal {} -> True
+               NotProvedGoal {} -> True
                _       -> False
 
-  go path gs =
+  go gs =
     case gs of
-      AtLoc pl _ gs1  ->
-        do pl' <- jsLoc pl
-           go ((pl', pl) : path) gs1
-
       Branch g1 g2 ->
         let (now,rest) = partition isGoal (flatBranch [g1,g2]) in
-          (++) <$> concatMapM (go path) now <*> concatMapM (go path) rest
+          (++) <$> concatMapM go now <*> concatMapM go rest
 
-      Goal asmps conc triv proved
-        | skipSuccessReports opts
-        , Proved _ <- proved
-        -> pure []
+      ProvedGoal asmps conc locs triv
+        | skipSuccessReports opts -> pure []
+        | otherwise -> jsProvedGoal locs asmps conc triv
 
+      NotProvedGoal asmps conc explain locs cex
         | skipIncompleteReports opts
-        , (SimError _ (ResourceExhausted _), _) <- conc
+        , SimError _ (ResourceExhausted _) <- conc
         -> pure []
 
-        | otherwise
-        -> do s <- jsSideCond apath asmps conc triv proved
-              pure [s]
+        | otherwise -> jsNotProvedGoal locs asmps conc explain cex
+
+
+removeRepeats :: Eq a => [a] -> [a]
+removeRepeats [] = []
+removeRepeats [x] = [x]
+removeRepeats (x:y:zs)
+  | x == y = removeRepeats (y:zs)
+  | otherwise = x : removeRepeats (y:zs)
+
+jsPath :: [ProgramLoc] -> IO [ JS ]
+jsPath locs = concat <$> mapM mkStep locs'
+    where
+      locs'      = annotateLoops (removeRepeats locs)
+      mkStep (a,l) =
+       jsLoc l >>= \case
+         Nothing -> return []
+         Just l' ->
+           return [jsObj
+             [ "loop" ~> jsList (map jsNum a)
+             , "loc"  ~> l'
+             ]]
+
+jsProvedGoal ::
+  [ ProgramLoc ] ->
+  [ CrucibleAssumption (Const ()) ] ->
```

Comment:
`Const ()` is a rather peculiar type. I wonder if it would be more direct to replace it with a data type with a nullary constructor and a single phantom type parameter, such as `Proxy`.

Reply:
I'm not opposed to `Proxy` if you think that's clearer.

Reply:
Upon a closer inspection, `Const ()` is already used in a number of places in `crucible-concurrency`, `crux-mir`, and `uc-crux-llvm`. I'm fine with sticking to the existing convention.